### PR TITLE
Customize UA strings

### PR DESCRIPTION
**Pull request by simonhong**
Brave/+ chromium major version string(ex Brave/74) is appended to chromium ua.
To test with original UA, --disable-override-ua switch is introduced.

As we use custom UA, UA whitelist and referral header are not used anymore.

Fix brave/brave-browser#1052

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Windows
  - [x] macOS
  - [ ] Linux
- Verified that these changes pass automated tests (`npm test brave_unit_tests && npm test brave_browser_tests && npm run test-security`) on
  - [ ] Windows
  - [x] macOS
  - [ ] Linux
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [x] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection

## Test Plan:
`yarn test brave_browser_tests --filter=BraveContentBrowserClientTest.OverrideUATest`
`yarn test brave_browser_tests --filter=BraveContentBrowserClientWithoutUAOverride.DisableOverrideUATest`

Manual test 1.
* Launch browser and open inspector
* Check "Brave/XX" is appended to ua by checking `navigator.userAgent`

Manual test 2.
* Launch browser with `--disable-override-ua` switch
* Check "Brave/XX" is *not* appended to ua by checking `navigator.userAgent`

Manual test 3.
* Launch browser and loading cheddar.com
* Check `x-brave-partner` is not used.

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [x] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: components/brave_referrals/browser/brave_referrals_service.cc [hunk 1]
```diff
@@ -100,208 +58,30 @@ void BraveReferralsService::Start() {
...
-        ProfileManager::GetLastUsedProfile());
-              content::OpenURLParams open_url_params(gurl, content::Referrer(),
-        WindowOpenDisposition::NEW_FOREGROUND_TAB,
-        ui::PAGE_TRANSITION_AUTO_TOPLEVEL, false);
-    open_url_params.extra_headers = FormatExtraHeaders(headers, gurl);
-    browser_displayer.browser()->OpenURL(open_url_params);
-  }
-
-  task_runner_->PostTask(FROM_HERE,
-                         base::Bind(&BraveReferralsService::DeletePromoCodeFile,
```
Comment (simonhong) [inline components/brave_referrals/browser/brave_referrals_service.cc]: @emerick How to handle this `DeletePromoCodeFile()` after deleting this callback?

File: browser/importer/brave_profile_writer.cc [hunk 2]
```diff
@@ -350,23 +350,6 @@ void BraveProfileWriter::UpdateReferral(const BraveReferral& referral) {
   } else {
     local_state->ClearPref(kReferralPromoCode);
   }
-
```
Comment (emerick) [inline browser/importer/brave_profile_writer.cc]: I don't think we want to delete this block, as this is just related to referral codes and not the headers.

File: common/pref_names.cc [hunk 3]
```diff
@@ -30,12 +30,6 @@ const char kUseOverriddenBraveThemeType[] =
     "brave.theme.use_overridden_brave_theme_type";
 const char kLocationBarIsWide[] = "brave.location_bar_is_wide";
 const char kReferralPromoCode[] = "brave.referral.promo_code";
-const char kReferralDownloadID[] = "brave.referral.download_id";
```
Comment (emerick) [inline common/pref_names.cc]: I believe we should keep all of these except kReferralHeaders.

File: common/pref_names.h [hunk 4]
```diff
@@ -27,11 +27,6 @@ extern const char kBraveThemeType[];
 extern const char kUseOverriddenBraveThemeType[];
 extern const char kLocationBarIsWide[];
 extern const char kReferralPromoCode[];
-extern const char kReferralDownloadID[];
```
Comment (emerick) [inline common/pref_names.h]: Same as above, keep all but kReferralHeaders.

File: components/brave_referrals/browser/brave_referrals_service.cc [hunk 5]
```diff
@@ -27,57 +26,16 @@
...
-const int kFetchReferralHeadersFrequency = 60 * 60 * 24;
-
-// Maximum size of the referral server response in bytes.
-const int kMaxReferralServerResponseSizeBytes = 1024 * 1024;
 
 // Default promo code, used when no promoCode file exists on first
 // run.
 const char kDefaultPromoCode[] = "BRV001";
 
-namespace {
```
Comment (emerick) [inline components/brave_referrals/browser/brave_referrals_service.cc]: Need to keep this as it contains the code for initializing the referral and sending the platform identifier for stats reporting.

File: components/brave_referrals/browser/brave_referrals_service.cc [hunk 6]
```diff
@@ -100,208 +58,30 @@ void BraveReferralsService::Start() {
...
-      base::TimeDelta::FromSeconds(kFetchReferralHeadersFrequency +
-                                   base::RandInt(0, 60 * 10)),
-      this, &BraveReferralsService::OnFetchReferralHeadersTimerFired);
-  DCHECK(fetch_referral_headers_timer_->IsRunning());
-
   // On first run, read the promo code from user-data-dir and
   // initialize the referral.
   bool checked_for_promo_code_file =
       pref_service_->GetBoolean(kReferralCheckedForPromoCodeFile);
-  std::string download_id = pref_service_->GetString(kReferralDownloadID);
```
Comment (emerick) [inline components/brave_referrals/browser/brave_referrals_service.cc]: Any references to download_id should stay, as that's how we track the referral in the browser.

File: components/brave_referrals/browser/brave_referrals_service.cc [hunk 7]
```diff
@@ -100,208 +58,30 @@ void BraveReferralsService::Start() {
...
-  base::Optional<base::Value> root =
-      base::JSONReader().ReadToValue(*response_body);
-  if (!root || !root->is_list()) {
-    LOG(ERROR) << "Failed to parse referral headers response";
-    return;
-  }
-  pref_service_->Set(kReferralHeaders, root.value());
-}
-
-void BraveReferralsService::OnReferralInitLoadComplete(
```
Comment (emerick) [inline components/brave_referrals/browser/brave_referrals_service.cc]: OnReferralInitLoadComplete should remain as that contains the functionality to handle successful referral initialization.

File: components/brave_referrals/browser/brave_referrals_service.cc (hunk 1 above)
Comment (emerick) [reply components/brave_referrals/browser/brave_referrals_service.cc]: We need to keep this callback, see above. :-)

File: components/brave_referrals/browser/brave_referrals_service.cc [hunk 8]
```diff
@@ -100,208 +58,30 @@ void BraveReferralsService::Start() {
...
-    open_url_params.extra_headers = FormatExtraHeaders(headers, gurl);
-    browser_displayer.browser()->OpenURL(open_url_params);
-  }
-
-  task_runner_->PostTask(FROM_HERE,
-                         base::Bind(&BraveReferralsService::DeletePromoCodeFile,
-                                    base::Unretained(this)));
-}
-
-void BraveReferralsService::OnReferralFinalizationCheckLoadComplete(
```
Comment (emerick) [inline components/brave_referrals/browser/brave_referrals_service.cc]: OnReferralFinalizationCheckLoadComplete should remain, as this is how we finalize the referral and discard our referral tracking info after 90 days which is very important.

File: components/brave_referrals/browser/brave_referrals_service.cc [hunk 9]
```diff
@@ -100,208 +58,30 @@ void BraveReferralsService::Start() {
...
-  pref_service_->ClearPref(kReferralAttemptTimestamp);
-  pref_service_->ClearPref(kReferralAttemptCount);
-}
-
 void BraveReferralsService::OnReadPromoCodeComplete() {
   pref_service_->SetBoolean(kReferralCheckedForPromoCodeFile, true);
-  if (!promo_code_.empty()) {
+  if (!promo_code_.empty())
     pref_service_->SetString(kReferralPromoCode, promo_code_);
-    InitReferral();
```
Comment (emerick) [inline components/brave_referrals/browser/brave_referrals_service.cc]: Call to `InitReferral` should remain here.

File: components/brave_referrals/browser/brave_referrals_service.cc [hunk 10]
```diff
@@ -314,12 +94,6 @@ void BraveReferralsService::GetFirstRunTime() {
       FROM_HERE, {content::BrowserThread::UI},
       base::BindOnce(&BraveReferralsService::MaybeDeletePromoCodePref,
                      base::Unretained(this)));
-
-  // Check for referral finalization, if appropriate.
```
Comment (emerick) [inline components/brave_referrals/browser/brave_referrals_service.cc]: This block should remain.

File: components/brave_referrals/browser/brave_referrals_service.cc [hunk 11]
```diff
@@ -356,47 +130,6 @@ void BraveReferralsService::DeletePromoCodeFile() const {
   }
 }
 
-void BraveReferralsService::MaybeCheckForReferralFinalization() {
```
Comment (emerick) [inline components/brave_referrals/browser/brave_referrals_service.cc]: `MaybeCheckForReferralFinalization` should remain.

File: components/brave_referrals/browser/brave_referrals_service.cc [hunk 12]
```diff
@@ -412,196 +145,6 @@ void BraveReferralsService::MaybeDeletePromoCodePref() const {
     pref_service_->ClearPref(kReferralPromoCode);
 }
 
-std::string BraveReferralsService::BuildReferralInitPayload() const {
```
Comment (emerick) [inline components/brave_referrals/browser/brave_referrals_service.cc]: `BuildReferralInitPayload` and `BuildReferralFinalizationCheckPayload` should remain. `FetchReferralHeaders` should go. `InitReferral`, `CheckForReferralFinalization`, etc. should remain. Basically, we only want to get rid of functionality that's directly tied to the referral headers.

File: components/brave_referrals/browser/brave_referrals_service.cc [hunk 13]
```diff
@@ -612,11 +155,6 @@ std::unique_ptr<BraveReferralsService> BraveReferralsServiceFactory(
 void RegisterPrefsForBraveReferralsService(PrefRegistrySimple* registry) {
   registry->RegisterBooleanPref(kReferralCheckedForPromoCodeFile, false);
   registry->RegisterStringPref(kReferralPromoCode, std::string());
-  registry->RegisterStringPref(kReferralDownloadID, std::string());
```
Comment (emerick) [inline components/brave_referrals/browser/brave_referrals_service.cc]: Only the `kReferralHeaders` call should be deleted here.

File: components/brave_referrals/browser/brave_referrals_service.cc [hunk 14]
```diff
@@ -100,208 +58,30 @@ void BraveReferralsService::Start() {
...
-
-  const base::Value* offer_page_url = root->FindKey("offer_page_url");
-  if (offer_page_url) {
-    GURL gurl(offer_page_url->GetString());
-    chrome::ScopedTabbedBrowserDisplayer browser_displayer(
-        ProfileManager::GetLastUsedProfile());
-              content::OpenURLParams open_url_params(gurl, content::Referrer(),
-        WindowOpenDisposition::NEW_FOREGROUND_TAB,
-        ui::PAGE_TRANSITION_AUTO_TOPLEVEL, false);
-    open_url_params.extra_headers = FormatExtraHeaders(headers, gurl);
```
Comment (emerick) [inline components/brave_referrals/browser/brave_referrals_service.cc]: Although this function should remain, this line that sets `extra_headers` should go.

File: common/network_constants.cc [hunk 15]
```diff
@@ -53,7 +53,6 @@ const char kTwitterRedirectURL[] = "https://mobile.twitter.com/i/nojs_router*";
 const char kCookieHeader[] = "Cookie";
 // Intentional misspelling on referrer to match HTTP spec
 const char kRefererHeader[] = "Referer";
-const char kUserAgentHeader[] = "User-Agent";
 const char kBravePartnerHeader[] = "X-Brave-Partner";
```
Comment (emerick) [inline common/network_constants.cc]: Can this be deleted?

**Comment by emerick** [comment]
cc: @aekeus: Just an FYI that we'll no longer be setting the `X-Brave-Partner` custom header with this change to the User-Agent string. One thing I hadn't considered is that since we won't be setting this header any more, I guess we no longer need to download the referral headers at all right? As far as I know, the only purpose that those headers serve is to tell us for which domains to set `X-Brave-Partner`. Please tell me if I'm missing something there, though.

File: common/network_constants.cc (hunk 15 above)
Comment (simonhong) [reply common/network_constants.cc]: I think so. this is only used by UA whitelist.

Comment (emerick) [reply common/network_constants.cc]: Sorry, just to be clear I'm talking about `kBravePartnerHeader`.

Comment (simonhong) [reply common/network_constants.cc]: I think it also can be deleted because it is only used by `OnBeforeStartTransaction_ReferralsWork()`.

**Comment by simonhong** [comment]
@emerick I think I resolved all things you commented. PTAL again! Thanks :)

File: browser/brave_content_browser_client.cc [hunk 16]
```diff
@@ -38,6 +39,7 @@
 #include "chrome/common/url_constants.h"
 #include "components/content_settings/core/browser/cookie_settings.h"
 #include "components/services/heap_profiling/public/mojom/heap_profiling_client.mojom.h"
+#include "components/version_info/version_info.h"
```
Comment (bridiver) [inline browser/brave_content_browser_client.cc]: is there already a dep for this in BUILD.gn?

Comment (simonhong) [reply browser/brave_content_browser_client.cc]: Oh, thanks for checking! Added.

File: browser/net/brave_site_hacks_network_delegate_helper.cc [hunk 17]
```diff
@@ -98,13 +100,6 @@ int OnBeforeStartTransaction_SiteHacksWork(net::URLRequest* request,
   if (IsBlockTwitterSiteHack(request, headers)) {
     return net::ERR_ABORTED;
   }
-  if (IsUAWhitelisted(request->url())) {
```
Comment (bridiver) [inline browser/net/brave_site_hacks_network_delegate_helper.cc]: this method doesn't appear to be used anywhere else so can be deleted if we decide we don't need this

Comment (simonhong) [reply browser/net/brave_site_hacks_network_delegate_helper.cc]: Deleted. Thanks!

File: components/brave_referrals/browser/brave_referrals_service.h [hunk 18]
```diff
@@ -49,8 +51,6 @@ class BraveReferralsService {
   std::string BuildReferralFinalizationCheckPayload() const;
   void FetchReferralHeaders();
```
Comment (emerick) [inline components/brave_referrals/browser/brave_referrals_service.h]: We shouldn't need `FetchReferralHeaders` now, right?

File: components/brave_referrals/browser/brave_referrals_service.h [hunk 19]
```diff
@@ -49,8 +51,6 @@ class BraveReferralsService {
...
   void CheckForReferralFinalization();
-  std::string FormatExtraHeaders(const base::Value* referral_headers,
-                                 const GURL& url);
 
   // Invoked from RepeatingTimer when referral headers timer fires.
   void OnFetchReferralHeadersTimerFired();
```
Comment (emerick) [inline components/brave_referrals/browser/brave_referrals_service.h]: We shouldn't need `OnFetchReferralHeadersTimerFired` now, right?

**Comment by emerick** [comment]
@simonhong This also updates `navigator.userAgent`, right? Maybe we should have a test for that.

**Comment by diracdeltas** [comment]
@emerick looks like there is a navigator.uA test here https://github.com/brave/brave-core/pull/2442/files#diff-df7253a767c4ecf0d635a48f9fc4aacdR540

**Comment by diracdeltas** [comment]
please update the documentation here as necessary: https://github.com/brave/brave-browser/wiki/Custom-Headers. this also needs an update to remove the item about UA: https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode#privacy-protection-enabled-regardless-of-whether-fingerprinting-protection-mode-is-on

File: components/brave_referrals/browser/brave_referrals_service.h (hunk 18 above)
Comment (simonhong) [reply components/brave_referrals/browser/brave_referrals_service.h]: Removed.

File: components/brave_referrals/browser/brave_referrals_service.h (hunk 19 above)
Comment (simonhong) [reply components/brave_referrals/browser/brave_referrals_service.h]: Removed.

**Comment by simonhong** [comment]
@emerick Regarding the `navigator.userAgent` change, I added some test cases for that.

**Comment by simonhong** [comment]
> please update the documentation here as necessary: https://github.com/brave/brave-browser/wiki/Custom-Headers (can just mention that as of this PR, there are no custom headers on desktop any more). this also needs an update to remove the item about UA: https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode#privacy-protection-enabled-regardless-of-whether-fingerprinting-protection-mode-is-on

Thanks for checking the documents. Will update when I get approve!

**Comment by simonhong** [comment]
Hmm, still can't be passed from Builder due to version mismatch even if I did rebase.
@bsclifton How can I handle this? Please help.

**Comment by bsclifton** [comment]
@simonhong quick heads up - @fmarier will be checking this out as part of the security review:
https://github.com/brave/internal/issues/571

**Comment by bsclifton** [comment]
I believe @petemill may have found a problem with this (when I had brought up https://github.com/brave/brave-browser/issues/4608)

Quick heads up @simonhong - the brave://version/ page does NOT show the `Brave/75` ending that is expected. When you're on the network tab and you inspect requests, it DOES show the proper header:
`Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.38 Safari/537.36 Brave/75`

**Comment by simonhong** [comment]
Hmm, `brave://version` page and `navigator.userAgent` api seems to use different api for fetching ua string..

**Comment by simonhong** [comment]
@bsclifton Created issue https://github.com/brave/brave-browser/issues/4623 for incomplete UA in `brave://version`

**Pull request merged**